src/impl_/pyclass.rs

```diff
@@ -1284,19 +1400,52 @@ impl<ClassT: PyClass, FieldT: ToPyObject, Offset: OffsetCalculator<ClassT, Field
 pub trait PyO3GetField: IntoPy<Py<PyAny>> + Clone {}
 impl<T: IntoPy<Py<PyAny>> + Clone> PyO3GetField for T {}
 
-/// Base case attempts to use IntoPy + Clone, which was the only behaviour before PyO3 0.22.
+/// IntoPy + Clone fallback case, which was the only behaviour before PyO3 0.22.
+impl<ClassT, FieldT, Offset>
+    PyClassGetterGenerator<ClassT, FieldT, Offset, false, false, true, false, false>
+where
+    ClassT: PyClass,
+    Offset: OffsetCalculator<ClassT, FieldT>,
+    FieldT: PyO3GetField,
+{
+    pub const fn generate(&self, name: &'static CStr, doc: &'static CStr) -> PyMethodDefType {
+        PyMethodDefType::Getter(PyGetterDef {
+            name,
+            meth: pyo3_get_value::<ClassT, FieldT, Offset>,
+            doc,
+        })
+    }
+}
+
+/// Base case attempts to use IntoPyObject + Clone
 impl<ClassT: PyClass, FieldT, Offset: OffsetCalculator<ClassT, FieldT>>
-    PyClassGetterGenerator<ClassT, FieldT, Offset, false, false>
+    PyClassGetterGenerator<ClassT, FieldT, Offset, false, false, false, false, false>
 {
+    #[cfg(not(diagnostic_namespace))]
     pub const fn generate(&self, name: &'static CStr, doc: &'static CStr) -> PyMethodDefType
     // The bound goes here rather than on the block so that this impl is always available
     // if no specialization is used instead
     where
-        FieldT: PyO3GetField,
+        for<'py> FieldT: IntoPyObject<'py> + Clone,
+        for<'py> PyErr: std::convert::From<<FieldT as IntoPyObject<'py>>::Error>,
     {
         PyMethodDefType::Getter(PyGetterDef {
             name,
-            meth: pyo3_get_value::<ClassT, FieldT, Offset>,
+            meth: pyo3_get_value_into_pyobject::<ClassT, FieldT, Offset>,
+            doc,
+        })
+    }
+
+    #[cfg(diagnostic_namespace)]
+    pub const fn generate(&self, name: &'static CStr, doc: &'static CStr) -> PyMethodDefType
+    // The bound goes here rather than on the block so that this impl is always available
+    // if no specialization is used instead
+    where
+        FieldT: PyO3GetFieldIntoPyObject,
```

Comment:
Maybe if the other `PyO3GetField` trait can be removed, this one could be renamed to just `PyO3GetField` instead.